Pipeline: output: posgresql: style (#1963)
fluent/fluent-bit-docs: pipeline/outputs/postgresql.md

# PostgreSQL

[PostgreSQL](https://www.postgresql.org) is an open source database management system that supports the SQL language and is capable of storing both structured and unstructured data, such as JSON objects.

Fluent Bit is designed to work with JSON objects, and the `pgsql` output plugin allows users to send their data to a PostgreSQL database and store it using the `JSONB` type.

PostgreSQL 9.4 or higher is required.

## Preliminary steps

According to the parameters set in the configuration file, the plugin creates the table defined by the `table` option in the database defined by the `database` option hosted on the server defined by the `host` option. It will use the PostgreSQL user defined by the `user` option, which must have the correct privileges to create such a table in that database.

{% hint style="info" %}

> If you're not familiar with how PostgreSQL's users and grants system works, the links in the [References](#references) section at might be helpful.

Reviewer: ```suggestion
If you're not familiar with how PostgreSQL's users and grants system works, the links in the [References](#references) section might be helpful.
```

If you prefer, instead of the `createuser` application, you can directly use the SQL command [`CREATE USER`](https://www.postgresql.org/docs/current/sql-createuser.html).

### Create the `fluentbit` database

As `postgres` system user, run:

```shell
createdb -O fluentbit fluentbit
```

This creates a database called `fluentbit` owned by the `fluentbit` user. As a result, the `fluentbit` user will be able to safely create the data table.

Alternatively, you can use the SQL command [`CREATE DATABASE`](https://www.postgresql.org/docs/current/sql-createdatabase.html).

### Connection

Ensure that the `fluentbit` user can connect to the `fluentbit` database on the specified target host. This might require you to properly configure the [`pg_hba.conf`](https://www.postgresql.org/docs/current/auth-pg-hba-conf.html) file.

## Configuration parameters

This plugin supports the following parameters:

| Key | Description | Default |
|:----|:------------|:-------|
| `Host` | Hostname/IP address of the PostgreSQL instance. | `- (127.0.0.1)`|
| `Port` | PostgreSQL port. | `- (5432)` |
| `User` | PostgreSQL username. | `- (current user)` |
| `Password` | Password of PostgreSQL username. | `-` |

> | `Database` | Database name to connect to.  | `- (current user)` |

Reviewer: ```suggestion
| `Database` | Database name to connect to. | `- (current user)` |
```

The output plugin automatically creates a table with the name specified by the `table` configuration option and made up of the following fields:

- `tag TEXT`
- `time TIMESTAMP WITHOUT TIMEZONE`
- `data JSONB`

The timestamp doesn't contain any information about the time zone, and it's therefore referred to the time zone used by the connection to PostgreSQL (`timezone` setting).

> For more information on the `JSONB` data type in PostgreSQL, refer to the [JSON types](https://www.postgresql.org/docs/current/datatype-json.html) page in the official documentation. You can find instructions on how to index or query the objects (including `jsonpath` introduced in PostgreSQL 12).

Reviewer: ```suggestion
For more information about the `JSONB` data type in PostgreSQL, refer to the [JSON types](https://www.postgresql.org/docs/current/datatype-json.html) page in the official documentation. You can find instructions on how to index or query the objects (including `jsonpath` introduced in PostgreSQL 12).
```

## Scalability

> PostgreSQL includes support for declarative partitioning. To improve vertical scalability of the database, you can partition your tables on time ranges, such as on a monthly basis. PostgreSQL supports also sub-partitions, allowing you to partition your records by hash and default partitions.

Reviewer: ```suggestion
PostgreSQL includes support for declarative partitioning. To improve vertical scalability of the database, you can partition your tables on time ranges, such as on a monthly basis. PostgreSQL supports also sub-partitions, which let you partition your records by hash and default partitions.
```

For more information on horizontal partitioning in PostgreSQL, refer to the [Table partitioning](https://www.postgresql.org/docs/current/ddl-partitioning.html) page in the official documentation.

Choose the latest major version of PostgreSQL if you're starting now.

## Additional information

PostgreSQL is a powerful and extensible database engine. More expert users can take advantage of `BEFORE INSERT` triggers on the main table and reroute records on normalised tables, depending on tags and content of the actual JSON objects.

> For example, you can use Fluent Bit to send HTTP log records to the `landing` table defined in the configuration file. This table contains a `BEFORE INSERT` trigger  (a function in `plpgsql` language) that normalizes the content of the JSON object and that inserts the record in another table (with its own structure and partitioning model). This kind of trigger lets you discard the record from the landing table by returning `NULL`.

Reviewer: ```suggestion
For example, you can use Fluent Bit to send HTTP log records to the `landing` table defined in the configuration file. This table contains a `BEFORE INSERT` trigger (a function in `plpgsql` language) that normalizes the content of the JSON object and that inserts the record in another table (with its own structure and partitioning model). This kind of trigger lets you discard the record from the landing table by returning `NULL`.
```

## References

> The following a list of resources from the PostgreSQL documentation:

Reviewer: ```suggestion
Refer to the following a list of resources from the PostgreSQL documentation:
```
